File: distance_matrix.py
from math import pi,sin,cos,atan2,sqrt
# import numpy as np

class Mat:

    content: list[list[float]]
    proximity: list[list[float]]
    size: int
    '''size: the height and length of the content in the matrix'''
    shape: list[int]
    beta: float

    def __init__(self, size:int, content:list[list[float]]=None) -> None:
        if(content):
            self.content = content
        else:
            self.content = [[0] * size for i in range(size)]
        # self.content = np.array((size,size))
        self.size = size
        self.shape = [size,size]
        # self.beta = beta
        # self.proximity = [[0] * size for i in range(size)]

    def set(self,i:int,j:int,distance:float) -> None:
        '''
            Replaces the value at position i,j with the distance parameter
        '''
        self.content[i][j] = distance

    def get(self,i:int,j:int) -> float:
        '''
            Returns the value held at position i,j
        '''
        return self.content[i][j]
    
    def all(self) -> list[list[float]]:
        '''
            Returns entire distance matrix
        '''
        return self.content

    def getProx(self,i:int,j:int) ->  float:
        '''
            Returns the proximity of the value at i,j rather than the distance
        '''
        return self.proximity[i][j]
    
    def row(self,i:int) -> list[float]:
        '''
            Returns the row of held at position i
        '''
        return self.content[i]
    
    def formDistMat(self,vertexCoords:list[list[float]],distance:str) -> None:
        '''
            Forms a distance matrix between all the given vertex coordinates
            vertexCoords: a list of coordinates for each data point vertex
            distance: the distance function to be used in the distance matrix calculations
        '''
        if(distance == "haversine"):
            distFunc = self.haversineDistance
        else:
            distFunc = self.pythagoreanDistance
        for i,n in enumerate(vertexCoords):
            for j,m in enumerate(vertexCoords):
                if(n==m):
                    continue
                self.content[i][j] = distFunc(n,m)

    def haversineDistance(self,i:list[float],j:list[float]) -> float:
        '''
            Use haversine forumla for finding distances between longitude and latitude coords
            i: first position
            j: second position
        '''
        R = 6371e3
        phiI = i[0] * pi / 180
        phiJ = j[0] * pi / 180
        deltaPhi = (j[0]-i[0]) * pi / 180
        deltaLambda = (j[1]-i[1]) * pi / 180

        a = (sin(deltaPhi/2) * sin(deltaPhi/2)) + (cos(phiI) * cos(phiJ) * sin(deltaLambda/2) * sin(deltaLambda/2))
        c = 2 * atan2(sqrt(a), sqrt(1-a))
        distance = R * c
        return distance

    def pythagoreanDistance(self,a:float,b:float) -> float:
        '''
            Use pythagoras formula to calculate distance between two points
        '''
        xDist = abs(a[0]-b[0])
        yDist = abs(a[1]-b[1])
        distance = sqrt((xDist**2) + (yDist**2))

        return distance

File: distancematrix.cpp
#include "distancematrix.h"
#include <math.h>
#include <iostream>
#include <fstream>
#include <sstream>

distanceMatrix::distanceMatrix(){

}

void distanceMatrix::loadFile(std::string file, int nodes)
{
    std::vector<std::vector<double>> coords = loadCSV(file,2,3,nodes,true);
    for (int i = 0; i < coords.size(); i++) {
        std::vector<double> newRow;
        for (int j = 0; j < coords.size(); j++) {
            if(i == j){
                newRow.push_back(0);
                continue;
            }
            double dist = haversine(coords[i], coords[j]);
            newRow.push_back(dist);
        }
        content.push_back(newRow);
    }
}

double distanceMatrix::get(int i, int j)
{
    return content[i][j];
}

void distanceMatrix::set(int i, int j, double val)
{
    content[i][j] = val;
}

std::vector<double> distanceMatrix::row(int i)
{
    return content[i];
}

std::vector<std::vector<double>> distanceMatrix::getAll()
{
    return content;
}

void distanceMatrix::print()
{
    for (int i = 0; i < content.size(); i++)
    {
        for (int j = 0; j < content[i].size(); j++)
        {
            std::cout << content[i][j] << " ";
        }
        std::cout<< "\n";
    }
}

double distanceMatrix::haversine(std::vector<double> i, std::vector<double> j)
{
    const double R = 6371e3;
    double phiI = i[0] * M_PI / 180;
    double phiJ = j[0] * M_PI / 180;
    double deltaPhi = (j[0]-i[0]) * M_PI / 180;
    double deltaLambda = (j[1]-i[1]) * M_PI / 180;

    double a = (sin(deltaPhi/2) * sin(deltaPhi/2)) + (cos(phiI) * cos(phiJ) * sin(deltaLambda/2) * sin(deltaLambda/2));
    double c = 2 * atan2(sqrt(a), sqrt(1-a));
    double distance = R * c;
    return distance;
}

std::vector<std::vector<double>> distanceMatrix::loadCSV(std::string file, int col1, int col2, int nodes, bool header)
{
    std::string line;

    std::ifstream fin(file);

    std::vector<std::vector<double>> coords;

    int nodesSoFar = 0;

    if(header)getline(fin,line);

    while (getline (fin, line) && nodesSoFar < nodes) {
        std::vector<double> coord;

        std::istringstream ss(line);
        std::string token;

        int count = 0;

        while(std::getline(ss, token, ',')) {
            count++;
            if(count == col1 or count == col2){
                coord.push_back(std::stod(token));
            }
            if(count > col2)break;
        }

        coords.push_back(coord);
        nodesSoFar++;
    }

    fin.close();

    return coords;
}
